File: src/main/kotlin/net/wavem/uvc/rms/gateway/event/response/EventResponseHandler.kt
package net.wavem.uvc.rms.gateway.event.response

import com.google.gson.Gson
import com.google.gson.JsonObject
import net.wavem.uvc.mqtt.application.MQTTService
import net.wavem.uvc.mqtt.domain.MQTTConnectionType
import net.wavem.uvc.rms.common.application.TimeService
import net.wavem.uvc.rms.common.application.UUIDService
import net.wavem.uvc.rms.common.domain.header.Header
import net.wavem.uvc.rms.common.domain.job.result.JobResult
import net.wavem.uvc.rms.common.types.area.AreaClsfType
import net.wavem.uvc.rms.common.types.event.ComInfoStatusType
import net.wavem.uvc.rms.common.types.event.EventCodeType
import net.wavem.uvc.rms.common.types.header.RobotType
import net.wavem.uvc.rms.common.types.job.JobGroupType
import net.wavem.uvc.rms.common.types.job.JobKindType
import net.wavem.uvc.rms.common.types.job.JobResultType
import net.wavem.uvc.rms.gateway.event.domain.Event
import net.wavem.uvc.rms.gateway.event.domain.EventProperties
import net.wavem.uvc.rms.gateway.event.domain.com_info.ComInfo
import net.wavem.uvc.rms.gateway.event.domain.event_info.EventInfo
import net.wavem.uvc.rms.gateway.event.domain.job.EventTaskInfo
import net.wavem.uvc.rms.gateway.event.dto.EventInfoDTO
import net.wavem.uvc.rms.gateway.location.domain.position.LocationPosition
import net.wavem.uvc.rms.gateway.location.dto.LocationPositionDTO
import net.wavem.uvc.ros.application.topic.Subscription
import net.wavem.uvc.ros.domain.robot_status_msgs.SensorStatus
import net.wavem.uvc.ros.domain.sensor_msgs.BatteryState
import net.wavem.uvc.ros.domain.sensor_msgs.NavSatFix
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class EventResponseHandler(
    private val eventProperties : EventProperties,
    private val mqttService : MQTTService<String>,
    private val timeService : TimeService,
    private val gson : Gson,
    private val uuidService : UUIDService
) {

    private val logger : Logger = LoggerFactory.getLogger(this.javaClass)

    private val rclGPSSubscription : Subscription<NavSatFix> = Subscription()
    private val rclBatteryStateSubscription : Subscription<BatteryState> = Subscription()
    private val rclIMUStatusSubscription : Subscription<SensorStatus> = Subscription()
    private val rclScanStatusSubscription : Subscription<SensorStatus> = Subscription()
    private val rclGPSStatusSubscription : Subscription<SensorStatus> = Subscription()
    private val rclBatteryStateStatusSubscription : Subscription<SensorStatus> = Subscription()

    init {
        val rclGpsSubscriptionTopic : String = "/ublox/fix"
        this.rclGPSSubscription.registerSubscription(rclGpsSubscriptionTopic, NavSatFix::class)
        this.logger.info("RCL {$rclGpsSubscriptionTopic} subscription registered")

        val rclBatteryStateSubscriptionTopic : String = "/battery_state"
        this.rclBatteryStateSubscription.registerSubscription(rclBatteryStateSubscriptionTopic, BatteryState::class)
        logger.info("RCL {$rclBatteryStateSubscriptionTopic} subscription registered")

        val rclIMUStatusSubscriptionTopic : String = "/imu/status"
        this.rclIMUStatusSubscription.registerSubscription(rclIMUStatusSubscriptionTopic, SensorStatus::class)
        this.logger.info("RCL {$rclIMUStatusSubscriptionTopic} subscription registered")

        val rclScanStatusSubscriptionTopic : String = "/scan/status"
        this.rclScanStatusSubscription.registerSubscription(rclScanStatusSubscriptionTopic, SensorStatus::class)
        this.logger.info("RCL {$rclScanStatusSubscriptionTopic} subscription registered")

        val rclGPSStatusSubscriptionTopic : String = "/ublox/fix/status"
        this.rclGPSStatusSubscription.registerSubscription(rclGPSStatusSubscriptionTopic, SensorStatus::class)
        this.logger.info("RCL {$rclGPSStatusSubscriptionTopic} subscription registered")

        val rclBatteryStateStatusSubscriptionTopic : String = "/battery_state/status"
        this.rclBatteryStateStatusSubscription.registerSubscription(rclBatteryStateSubscriptionTopic, SensorStatus::class)
        this.logger.info("RCL {$rclBatteryStateStatusSubscriptionTopic} subscription registered")
    }

    private fun buildHeader() : Header {
        return Header(
            robotCorpId = "rco0000001",
            workCorpId = "wco0000001",
            workSiteId = "wst0000001",
            robotId = "rbt0000001",
            robotType = RobotType.AMR.type
        )
    }

    private fun buildJobResult() : JobResult {
        this.rclBatteryStateSubscription.getDataObservable().subscribe {
            val batteryState : BatteryState = BatteryState.read(it)
            logger.info("RCL {/battery_state} subscription callback : $batteryState")
        }

        return JobResult(
            status = JobResultType.SUCCESS.type,
            startTime = timeService.getCurrentDateTime(),
            endTime = timeService.getCurrentDateTime(),
            startBatteryLevel = 70,
            endBatteryLevel = 65,
            dist = 300
        )
    }

    private fun buildJobInfo() : EventTaskInfo {
        var jobPlanId : String = ""
        val jobPlanIdFromYML : String? = System.getProperty("jobInfo.jobPlanId")
        if (jobPlanIdFromYML != null) {
            jobPlanId = jobPlanIdFromYML
        }

        var jobGroupId : String = ""
        val jobGroupIdFromYML : String? = System.getProperty("jobInfo.jobGroupId")
        if (jobGroupIdFromYML != null) {
            jobGroupId = jobGroupIdFromYML
        }

        var jobOrderId : String = ""
        val jobOrderIdFromYML : String? = System.getProperty("jobInfo.jobOrderId")
        if (jobOrderIdFromYML != null) {
            jobOrderId = jobOrderIdFromYML
        }

        return EventTaskInfo(
            jobPlanId = jobPlanId,
            jobGroupId = jobGroupId,
            jobOrderId = jobOrderId,
            jobGroup = JobGroupType.SUPPLY.type,
            jobKindType = JobKindType.MOVE.type,
            jobResult = this.buildJobResult()
        )
    }

    private fun buildLocationPosition() : LocationPosition {
        val locationPositionDTO : LocationPositionDTO = LocationPositionDTO()

        this.rclGPSSubscription.getDataObservable().subscribe {
            val navSatFix : NavSatFix = NavSatFix.read(it)
            logger.info("RCL {/ublox/fix} subscription callback : $navSatFix")
            locationPositionDTO.setXpos(navSatFix.latitude)
            locationPositionDTO.setYpos(navSatFix.longitude)
            locationPositionDTO.setHeading(45.0)
        }

        return locationPositionDTO.build()
    }

    private fun buildEventInfo() : EventInfo {
        val eventInfoDTO : EventInfoDTO = EventInfoDTO()
        eventInfoDTO.eventId = uuidService.generateUUID()
        eventInfoDTO.areaClsf = AreaClsfType.INDOOR.type
        eventInfoDTO.floor = "1F"

        this.rclIMUStatusSubscription.getDataObservable().subscribe {
            val imuStatus : SensorStatus = SensorStatus.read(it)

            if (imuStatus.status_code == -1000) {
                eventInfoDTO.eventCd = EventCodeType.BROKEN.type
                eventInfoDTO.eventSubCd = imuStatus.status_message
            }
        }

        this.rclScanStatusSubscription.getDataObservable().subscribe {
            val scanStatus : SensorStatus = SensorStatus.read(it)

            if (scanStatus.status_code == -1001) {
                eventInfoDTO.eventCd = EventCodeType.BROKEN.type
                eventInfoDTO.eventSubCd = scanStatus.status_message
            }
        }

        this.rclGPSStatusSubscription.getDataObservable().subscribe {
            val gpsStatus : SensorStatus = SensorStatus.read(it)

            if (gpsStatus.status_code == -1002) {
                eventInfoDTO.eventCd = EventCodeType.BROKEN.type
                eventInfoDTO.eventSubCd = gpsStatus.status_message
            }
        }

        this.rclBatteryStateStatusSubscription.getDataObservable().subscribe {
            val batteryStatus : SensorStatus = SensorStatus.read(it)

            if (batteryStatus.status_code == -1003) {
                eventInfoDTO.eventCd = EventCodeType.BROKEN.type
                eventInfoDTO.eventSubCd = batteryStatus.status_message
            }
        }

        this.rclBatteryStateSubscription.getDataObservable().subscribe {
            val batteryState : BatteryState = BatteryState.read(it)
            logger.info("RCL {/battery_state} subscription callback : $batteryState")
        }

        eventInfoDTO.location = this.buildLocationPosition()

        return eventInfoDTO.build()
    }

    private fun buildComInfo() : ComInfo {
        return ComInfo(
            status = ComInfoStatusType.CONNECTED.type,
            robotIP = "192.168.1.100",
            mqttIP = mqttService.readMQTTYML().get("ip").asString,
            mqttPort = mqttService.readMQTTYML().get("port").asString
        )
    }

    fun handle() {
        val event : Event = Event(
            header = this.buildHeader(),
            taskInfo = this.buildJobInfo(),
            eventInfo = this.buildEventInfo(),
            comInfo = this.buildComInfo()
        )

        val eventJson : JsonObject = gson.toJsonTree(event).asJsonObject
        mqttService.bridge(MQTTConnectionType.TO_RMS, topic = eventProperties.topic, eventJson.toString())
    }

    companion object {
        const val RCL_GPS_FIX_TOPIC : String = "/ublox/fix"
    }
}

File: src/rms/response/event/handler/event_response_handler.py
import rclpy
import json

from rclpy.node import Node
from rclpy.subscription import Subscription
from rclpy.qos import qos_profile_sensor_data
from rclpy.qos import qos_profile_system_default
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup

from sensor_msgs.msg import NavSatFix
from robot_status_msgs.msg import SensorStatus

from mqtt import broker
from rms.common.application.uuid_service import UUIDService
from rms.common.application.time_service import TimeService
from rms.common.domain.type.robot.robot_type import RobotType
from rms.common.domain.type.area.area_clsf_type import AreaCLSFType
from rms.common.domain.type.job.job_group_type import JobGroupType
from rms.common.domain.type.job.job_kind_type import JobKindType
from rms.common.domain.type.job.job_result_status_type import JobResultStatusType
from rms.response.event.domain.eventInfo.type.event_cd_type import EventCdType
from rms.response.event.domain.comInfo.type.com_info_status import ComInfoStatusType

from rms.common.domain.header import Header
from rms.response.event.domain.event import Event
from rms.response.event.domain.taskInfo.task_info import TaskInfo
from rms.response.event.domain.taskInfo.jobResult.job_result import JobResult
from rms.response.event.domain.eventInfo.event_info import EventInfo
from rms.response.event.domain.eventInfo.location.event_info_location import EventInfoLocation
from rms.response.event.domain.comInfo.com_info import ComInfo


class EventResponseHandler():
    rclpy_ublox_fix_subscription_topic: str = '/ublox/fix'
    rclpy_imu_status_subscription_topic: str = '/imu/status'
    rclpy_scan_status_subscriptin_topic: str = '/scan/status'
    rclpy_ublox_fix_status_subscription_topic: str = '/ublox/fix/status'
    rclpy_battery_state_subscription_topic: str = '/battery/status'
    
    mqtt_event_publisher_topic: str = 'hubilon/atcplus/ros/event'
    
    
    def __init__(self, rclpy_node: Node, mqtt_broker: broker.mqtt_broker) -> None:
        self.rclpy_node: Node = rclpy_node
        
        self.ublox_fix_subscription_cb_group: MutuallyExclusiveCallbackGroup = MutuallyExclusiveCallbackGroup()
        self.ublox_fix_subscription: Subscription = self.rclpy_node.create_subscription(
            msg_type = NavSatFix,
            topic = self.rclpy_ublox_fix_subscription_topic,
            callback = self.rclpy_ublox_fix_subscription_cb,
            qos_profile = qos_profile_sensor_data,
            callback_group = self.ublox_fix_subscription_cb_group
        )

        self.imu_status_subscription_cb_group: MutuallyExclusiveCallbackGroup = MutuallyExclusiveCallbackGroup()
        self.imu_status_subscription: Subscription = self.rclpy_node.create_subscription(
            msg_type = SensorStatus,
            topic = self.rclpy_imu_status_subscription_topic,
            callback = self.rclpy_imu_status_subscription_cb,
            qos_profile = qos_profile_system_default,
            callback_group = self.imu_status_subscription_cb_group
        )

        self.scan_status_subscription_cb_group: MutuallyExclusiveCallbackGroup = MutuallyExclusiveCallbackGroup()
        self.scan_status_subscription: Subscription = self.rclpy_node.create_subscription(
            msg_type = SensorStatus,
            topic = self.rclpy_scan_status_subscriptin_topic,
            callback = self.rclpy_scan_status_subscription_cb,
            qos_profile = qos_profile_system_default,
            callback_group = self.imu_status_subscription_cb_group
        )
        
        self.ublox_fix_status_subscription_cb_group: MutuallyExclusiveCallbackGroup = MutuallyExclusiveCallbackGroup()
        self.ublox_fix_status_subscription: Subscription = self.rclpy_node.create_subscription(
            msg_type = SensorStatus,
            topic = self.rclpy_ublox_fix_subscription_topic,
            callback = self.rclpy_ublox_fix_status_subscription_cb,
            qos_profile = qos_profile_system_default,
            callback_group = self.ublox_fix_status_subscription_cb_group
        )
        
        self.battery_state_status_subscription_cb_group: MutuallyExclusiveCallbackGroup = MutuallyExclusiveCallbackGroup()
        self.battery_state_status_subscription: Subscription = self.rclpy_node.create_subscription(
            msg_type = SensorStatus,
            topic = self.rclpy_battery_state_subscription_topic,
            callback = self.rclpy_battery_state_status_subscription_cb,
            qos_profile = qos_profile_system_default,
            callback_group = self.battery_state_status_subscription_cb_group
        )

        self.mqtt_broker: broker.mqtt_broker = mqtt_broker
        self.uuid_service: UUIDService = UUIDService()
        self.time_service: TimeService = TimeService()
        
        self.start_battery_level: int = 0
        self.end_battery_level: int = 0
        self.dist: int = 0
        self.sensor_status: str = ""
        self.location_xpos: float = 0.0
        self.location_ypos: float = 0.0
        self.heading: float = 45.0
    
    
    def rclpy_ublox_fix_subscription_cb(self, ublox_fix_cb: NavSatFix) -> None:
        self.rclpy_node.get_logger().info(
            'Event GPS cb\n\tlat : "%f"\n\tlon : "%f"\n\talt : "%f"' % (
                ublox_fix_cb.latitude,
                ublox_fix_cb.longitude,
                ublox_fix_cb.altitude
            )
        )
        
        self.location_xpos = ublox_fix_cb.latitude
        self.location_ypos = ublox_fix_cb.longitude
        self.heading = ublox_fix_cb.altitude
    
    
    def __report_sensor_broken_status__(self, sensor_name: str) -> None:
        self.sensor_status = EventCdType.BROKEN.value
        self.rclpy_node.get_logger().info('Event %s Sensor Broken "%s"' % (sensor_name, self.sensor_status))
        self.response_to_uvc()
    

    def rclpy_imu_status_subscription_cb(self, imu_status_cb: SensorStatus) -> None:
        status_code: int = imu_status_cb.status_code
        status_message: str = imu_status_cb.status_message
        
        self.rclpy_node.get_logger().info(
            'Event IMU Status\n\tcode : "%d"\n\tmessage : "%s"' % (
                status_code,
                status_message
            )
        )
        
        if (status_code == -1000):
            self.__report_sensor_broken_status__('IMU')
        else:
            return
        
    
    def rclpy_scan_status_subscription_cb(self, scan_status_cb: SensorStatus) -> None:
        status_code: int = scan_status_cb.status_code
        status_message: str = scan_status_cb.status_message
        
        self.rclpy_node.get_logger().info(
            'Event SCAN Status\n\tcode : "%d"\n\tmessage : "%s"' % (
                status_code,
                status_message
            )
        )
        
        if (status_code == -1001):
            self.__report_sensor_broken_status__('SCAN')
        else:
            return
        
    
    def rclpy_ublox_fix_status_subscription_cb(self, ublox_fix_status_cb: SensorStatus) -> None:
        status_code: int = ublox_fix_status_cb.status_code
        status_message: str = ublox_fix_status_cb.status_message
        
        self.rclpy_node.get_logger().info(
            'Event UbloxFix Status\n\tcode : "%d"\n\tmessage : "%s"' % (
                status_code,
                status_message
            )
        )
        
        if (status_code == -1002):
            self.__report_sensor_broken_status__('GPS')
        else:
            return
        
    
    def rclpy_battery_state_status_subscription_cb(self, battery_state_status: SensorStatus) -> None:
        status_code: int = battery_state_status.status_code
        status_message: str = battery_state_status.status_message
        
        self.rclpy_node.get_logger().info(
            'Event BatteryState Status\n\tcode : "%d"\n\tmessage : "%s"' % (
                status_code,
                status_message
            )
        )
        
        if (status_code == -1003):
            self.__report_sensor_broken_status__('BATTERY')
        else:
            return
        

    def build_event(self) -> Event:
        header: Header = self.__build_header__()
        task_info: TaskInfo = self.__build__task_info__()
        event_info: EventInfo = self.__build_event_info__()
        com_info: ComInfo = self.__build_com_info__()

        event: Event = Event(
            header = header.__dict__,
            taskInfo = task_info.__dict__,
            eventInfo = event_info.__dict__,
            comInfo = com_info.__dict__
        )
        
        return event

    
    def __build_header__(self) -> Header:
        header: Header = Header(
            robotCorpId = 'rco0000001',
            workCorpId = 'wco0000001',
            workSiteId = 'wst0000001',
            robotId = 'rbt0000001',
            robotType = RobotType.AMR.value
        )

        return header
    

    def __build__task_info__(self) -> TaskInfo:
        job_plan_id: str = self.uuid_service.generate_uuid()
        job_group_id: str = self.uuid_service.generate_uuid()
        job_order_id: str = self.uuid_service.generate_uuid()
        
        formatted_datetime: str = self.time_service.get_current_datetime()

        job_result : JobResult = JobResult(
            status = JobResultStatusType.SUCCESS.value,
            startTime = formatted_datetime,
            endTime = formatted_datetime,
            startBatteryLevel = self.start_battery_level,
            endBatteryLevel = self.end_battery_level,
            dist = 300
        )
        
        taskInfo: TaskInfo = TaskInfo(
            jobPlanId = job_plan_id,
            jobGroupId = job_group_id,
            jobOrderId = job_order_id,
            jobGroup = JobGroupType.SUPPLY.value,
            jobKind = JobKindType.MOVE.value,
            jobResult = job_result.__dict__
        )
        
        return taskInfo
    
    
    def __build_event_info__(self) -> EventInfo:
        event_location: EventInfoLocation = EventInfoLocation(
            xpos = self.location_xpos,
            ypos = self.location_ypos,
            heading = self.heading
        )
        
        event_id: str = self.uuid_service.generate_uuid()
        
        event_info: EventInfo = EventInfo(
            eventId = event_id,
            eventCd = self.sensor_status,
            eventSubCd = 'SUB001',
            areaClsf = AreaCLSFType.INDOOR.value,
            floor = '1F',
            batteryLevel = 50,
            location = event_location.__dict__
        )

        return event_info
    

    def __build_com_info__(self) -> ComInfo:
        com_info: ComInfo = ComInfo(
            status = ComInfoStatusType.CONNECTED.value,
            robotIP = '192.168.0.1',
            mqttIP = '33',
            mqttPort = '1883'
        )

        return com_info


    def response_to_uvc(self) -> None:
        built_event: Event = self.build_event()
        self.mqtt_broker.client.publish(self.mqtt_event_publisher_topic, json.dumps(built_event.__dict__))
        

__all__ = ['event_response_handler']